sharding/contracts/sharding_manager_test.go

```diff
@@ -1,130 +1,560 @@
+	notaryDepositInsufficient, _ = new(big.Int).SetString("999000000000000000000", 10)
+	notaryDeposit, _             = new(big.Int).SetString("1000000000000000000000", 10)
+	smcConfig                    = SMCConfig{
+		notaryLockupLenght:   new(big.Int).SetInt64(1),
+		proposerLockupLength: new(big.Int).SetInt64(1),
+	}
+	ctx = context.Background()
 )
 
-func deploySMCContract(backend *backends.SimulatedBackend) (common.Address, *types.Transaction, *SMC, error) {
+func deploySMCContract(backend *backends.SimulatedBackend, key *ecdsa.PrivateKey) (common.Address, *types.Transaction, *SMC, error) {
 	transactOpts := bind.NewKeyedTransactor(key)
 	defer backend.Commit()
 	return DeploySMC(transactOpts, backend)
 }
 
 // Test creating the SMC contract
 func TestContractCreation(t *testing.T) {
-	backend := backends.NewSimulatedBackend(core.GenesisAlloc{addr: {Balance: accountBalance1001Eth}})
-	_, _, _, err := deploySMCContract(backend)
+	addr := crypto.PubkeyToAddress(mainKey.PublicKey)
+	backend := backends.NewSimulatedBackend(core.GenesisAlloc{addr: {Balance: accountBalance2000Eth}})
+	_, _, _, err := deploySMCContract(backend, mainKey)
 	backend.Commit()
 	if err != nil {
 		t.Fatalf("can't deploy SMC: %v", err)
 	}
 }
 
-// Test getting the collation gas limit
-func TestGetCollationGasLimit(t *testing.T) {
-	backend := backends.NewSimulatedBackend(core.GenesisAlloc{addr: {Balance: accountBalance1001Eth}})
-	_, _, smc, _ := deploySMCContract(backend)
-	gasLimit, err := smc.GetCollationGasLimit(&bind.CallOpts{})
+func TestNotaryRegister(t *testing.T) {
+	const notaryCount = 3
+	var notaryPoolAddr [notaryCount]common.Address
+	var notaryPoolPrivKeys [notaryCount]*ecdsa.PrivateKey
+	var txOpts [notaryCount]*bind.TransactOpts
+	genesis := make(core.GenesisAlloc)
+
+	// initializes back end with 3 accounts and each with 2000 eth balances
+	for i := 0; i < notaryCount; i++ {
+		key, _ := crypto.GenerateKey()
+		notaryPoolPrivKeys[i] = key
+		notaryPoolAddr[i] = crypto.PubkeyToAddress(key.PublicKey)
+		txOpts[i] = bind.NewKeyedTransactor(key)
+		txOpts[i].Value = notaryDeposit
+		genesis[notaryPoolAddr[i]] = core.GenesisAccount{
+			Balance: accountBalance2000Eth,
+		}
+	}
+
+	backend := backends.NewSimulatedBackend(genesis)
+	_, _, smc, _ := deploySMCContract(backend, notaryPoolPrivKeys[0])
+
+	// Test notary 0 has not registered
+	notary, err := smc.NotaryRegistry(&bind.CallOpts{}, notaryPoolAddr[0])
+	if err != nil {
+		t.Fatalf("Can't get notary registry info: %v", err)
+	}
+
+	if notary.Deposited {
+		t.Fatalf("Notary has not registered. Got deposited flag: %v", notary.Deposited)
+	}
+
+	// Test notary 0 has registered
+	if _, err := smc.RegisterNotary(txOpts[0]); err != nil {
+		t.Fatalf("Registering notary has failed: %v", err)
+	}
+	backend.Commit()
+
+	notary, err = smc.NotaryRegistry(&bind.CallOpts{}, notaryPoolAddr[0])
+
+	if !notary.Deposited &&
+		notary.PoolIndex.Cmp(big.NewInt(0)) != 0 &&
```

Comment:
Shouldn't these be `||`? Otherwise this will only fail if all of your conditions are met, but I think you meant to fail if any of them are met.

Reply:
Ah! good catch. It should be ||